src/compiler/checker.ts

```diff
@@ -4363,6 +4734,40 @@ namespace ts {
             return links.resolvedType;
         }
 
+        function getTypeFromJSDocFunctionType(node: JSDocFunctionType): Type {

```

Comment:
All of the `getTypeFromXXX` functions cache the type in the `resolvedType` property of `NodeLinks`, however the `getTypeFromJSDocXXX` functions don't. That means we'll generate new type identities every time a type is fetched  which is a waste of time and memory, plus it generates more work down stream when types are compared (because we appear to have different types that need structural comparisons).